common/persistence/sql/sqlplugin/postgres/task.go

```diff
@@ -48,7 +48,8 @@ task_type = :task_type
 
 	listTaskListQry = `SELECT domain_id, range_id, name, task_type, data, data_encoding ` +
 		`FROM task_lists ` +
-		`WHERE shard_id = $1 AND domain_id > $2 AND name > $3 AND task_type > $4 ORDER BY domain_id,name,task_type LIMIT $5`
+		`WHERE shard_id = $1 AND ((domain_id = $2 AND name = $3 AND task_type > $4) OR (domain_id=$2 AND name > $3) OR (domain_id > $2)) ` +
+		`ORDER BY domain_id,task_type,name LIMIT $5`
 
```

Comment:
Trying to understand what's the problem and how are you solving it. Here is my learning, please correct me if I am wrong. 
The goal of `ListTasklists` persistence API is to go through every tasklists. Implicitly, there are the requirements:
* Cannot miss any tasklist, cannot return same tasklist more than once
* Must aways move forward, given whatever pagesize. 

Basically we are using PK to implement cursor based pagination(because it's more efficient/scalable than SQL's LIMIT+Offset).


The code here is working tightly with sqlTaskManager's `ListTaskList`. :
sqlTaskManager is relying on PaginationToken to record the status. sqlTaskManager will start with `shardID=0` and loop until `maxShards-1`. 
The contract between the two sqlTaskManger and here(task.go) is that sqlTaskManager will pass in the last record as the cursor. 
For each shard, sqlTaskManager will start with this fake record as cursor:
`taskType=math.MinInt, domainID=minUUID`. And then moving on, it will use last returned record in the result set.
```
		lastRow := &rows[request.PageSize-1]
		nextPageToken, err = gobSerialize(&taskListPageToken{
			ShardID:  pageToken.ShardID,
			DomainID: lastRow.DomainID.String(),
			Name:     lastRow.Name,
			TaskType: lastRow.TaskType,
		})
```

Hence the contract for task.go is to move forward(exclusively) from this last record. 

The existing one is apparently broken, because for example -- 
`domain_id > $2 AND name > $3 AND task_type > $4`, domain_id > $2 would have to skip some records, if the last record's domain still have some other records(tasklists). 

Therefor you are doing:
`((domain_id = $2 AND name = $3 AND task_type > $4) OR (domain_id=$2 AND name > $3) OR (domain_id > $2))`
This will ensure:
* If last record's domain still have other tasklists, then will try to get from them, but they must be greater than last tasklist
* If Not, then get next domain_id( domain_id > $2 )

Reply:
Thanks for doing it. Took me a little while to understand. 

btw I was thinking to point you out to another way that I implemented cursor based pagination for our event history store( see https://github.com/uber/cadence/blob/bfcb03fe53883ff4d574d9cc41e20dab202d4d52/common/persistence/sql/sqlHistoryManager.go#L146), but now I believe your change is smaller and safer to do. 

Would you mind help us add some comments to the code? If my above understanding is correct, you may just copy /modify most of them, hehe. Maybe it's better put in sqlTaskManager(like we added comment for sqlHistoryManager), as it has the most logic. 
Comment will help us maintain the code much easier. As people can understand it now, but they will forget it a few months later.